Softeq.XToolkit.Common/Commands/AsyncCommandGeneric.cs

```diff
@@ -17,104 +19,149 @@ namespace Softeq.XToolkit.Common.Commands
+        ///     to be passed, this object can be set to a null reference.
         /// </param>
-        /// <param name="canExecute">The Function that verifies whether or not AsyncCommand should execute.</param>
-        /// <param name="onException">If an exception is thrown in the Task, <c>onException</c> will execute.</param>
-        public AsyncCommand(
-            Func<T, Task> execute,
-            Func<object?, bool>? canExecute,
-            Action<Exception>? onException = null)
-            : base(canExecute)
-        {
-            _execute = execute;
-            _onException = onException;
-        }
-
-        /// <inheritdoc cref="AsyncCommandBase.CanExecute"/>
-        public bool CanExecute(T parameter)
+        /// <returns>true if this command can be executed; otherwise, false.</returns>
+        public bool CanExecute(object? parameter)
         {
-            return CanExecute(parameter as object);
+            return TryParseParameter(parameter, out T parsed) && CanExecute(parsed);
         }
 
         /// <inheritdoc cref="AsyncCommand.Execute"/>
         public void Execute(object? parameter)
         {
-            switch (parameter)
+            if (TryParseParameter(parameter, out T parsed))
             {
-                case T validParameter:
-                    Execute(validParameter);
-                    break;
-
-                case null when !typeof(T).GetTypeInfo().IsValueType:
-                    Execute(default!);
-                    break;
-
-                case null:
-                    throw new InvalidCommandParameterException(typeof(T));
-
-                default:
-                    throw new InvalidCommandParameterException(typeof(T), parameter.GetType());
+                Execute(parsed);
+            }
+            else
+            {
+                AssertParameterNotSupported(parameter);
             }
         }
 
         /// <inheritdoc cref="AsyncCommand.Execute"/>
         /// <typeparam name="T">Type of parameter.</typeparam>
         public void Execute(T parameter)
         {
-            ExecuteAsync(parameter).FireAndForget(_onException!);
+            ExecuteAsync(parameter).FireAndForget(TryHandleException);
         }
 
+        /// <inheritdoc cref="IAsyncCommand.ExecuteAsync"/>
+        /// <typeparam name="T">Type of parameter.</typeparam>
         public Task ExecuteAsync(T parameter)
         {
-            return ExecuteAsync(_execute, parameter);
+            return CanExecute(parameter)
+                ? DoExecuteAsync(() => _execute.Execute(parameter))
+                : Task.CompletedTask;
+        }
+
+        private static bool TryParseParameter(object? parameter, out T parsed)
+        {
+            switch (parameter)
+            {
+                case T p:
+                    parsed = p;
+                    return true;
+                case null when !typeof(T).GetTypeInfo().IsValueType:
+                    parsed = default!;
```

Comment:
Just wondering why compiler doesn't like it without '!' after 'default'? We will definitely return null in this case but we tell compiler that it won't be null.